src/main/java/net/minecraftforge/server/command/TagsCommand.java

```diff
@@ -0,0 +1,266 @@
+        final long totalPages = (count - 1) / PAGE_SIZE + 1;
+        final long actualPage = Mth.clamp(currentPage, 1, totalPages);
+
+        MutableComponent containsComponent = new TranslatableComponent(containsText, count);
+        if (count > 0) // Highlight the count text, make it clickable, and append page counters
+        {
+            containsComponent = ComponentUtils.wrapInSquareBrackets(containsComponent.withStyle(s -> s
+                    .withColor(ChatFormatting.GREEN)
+                    .withClickEvent(new ClickEvent(ClickEvent.Action.COPY_TO_CLIPBOARD, allElementNames))
+                    .withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT,
+                            new TranslatableComponent(copyHoverText)))));
+            containsComponent = new TranslatableComponent("%s <page %s / %s>",
+                    containsComponent, actualPage, totalPages);
+        }
+
+        final MutableComponent tagElements = new TextComponent("").append(containsComponent);
+        names.get()
+                .sorted()
+                .skip(PAGE_SIZE * (actualPage - 1))
+                .limit(PAGE_SIZE)
+                .map(TextComponent::new)
+                .map(t -> t.withStyle(elementColor))
+                .map(t -> new TranslatableComponent("\n - %s", t))
+                .forEach(tagElements::append);
+
+        return header.append("\n").append(tagElements);
+    }
+
+    private static CompletableFuture<Suggestions> suggestRegistries(final CommandContext<CommandSourceStack> ctx,
+            final SuggestionsBuilder builder)
+    {
+        ctx.getSource().suggestRegistryElements(Registry.REGISTRY,
+                SharedSuggestionProvider.ElementSuggestionType.ELEMENTS, builder);
+        return builder.buildFuture();
+    }
+
+    private static SuggestionProvider<CommandSourceStack> suggestFromRegistry(
+            final Function<Registry<?>, Iterable<ResourceLocation>> namesFunction)
+    {
+        return (ctx, builder) -> getResourceKey(ctx, "registry", ROOT_REGISTRY_KEY)
+                .flatMap(key -> ctx.getSource().registryAccess().registry(key).map(registry -> {
+                    SharedSuggestionProvider.suggestResource(namesFunction.apply(registry), builder);
+                    return builder.buildFuture();
+                }))
+                .orElseGet(builder::buildFuture);
+    }
+
+    @SuppressWarnings("SameParameterValue")
+    private static <T> Optional<ResourceKey<T>> getResourceKey(final CommandContext<CommandSourceStack> ctx,
+            final String name,
+            final ResourceKey<Registry<T>> registryKey)
+    {
+        // Don't inline to avoid an unchecked cast warning due to raw types
+        final ResourceKey<?> key = ctx.getArgument(name, ResourceKey.class);
+        return key.cast(registryKey);
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <I, O> O cast(final I input)
```

Comment:
Is the `I` being a generic actually needed? Or can this just be `private static <O> cast(final Object input)`